src/common/m_boundary_common.fpp

```diff
@@ -0,0 +1,582 @@
+
+!> @brief The purpose of the module is to apply noncharacteristic and processor
+!! boundary condiitons
+
+#:include 'macros.fpp'
+#:include 'inline_boundary_conditions.fpp'
+
+module m_boundary_common
+
+    use m_derived_types        !< Definitions of the derived types
+
+    use m_global_parameters    !< Definitions of the global parameters
+
+    use m_mpi_proxy
+
+    use m_constants
+
+    implicit none
+
+    type(scalar_field), dimension(:, :), allocatable :: bc_buffers
+    !$acc declare create(bc_buffers)
+
+    real(wp) :: bcxb, bcxe, bcyb, bcye, bczb, bcze
+
+#ifdef MFC_MPI
+    integer, dimension(1:3, -1:1) :: MPI_BC_TYPE_TYPE, MPI_BC_BUFFER_TYPE
+#endif
+
+#ifdef MFC_SIMULATION
+    private; public :: s_initialize_boundary_common_module, &
+ s_populate_variables_buffers, &
+ s_create_mpi_types, &
+ s_populate_capillary_buffers, &
+ s_finalize_boundary_common_module
+#else
+    private; public :: s_initialize_boundary_common_module, &
+ s_populate_variables_buffers, &
+ s_create_mpi_types, &
+ s_finalize_boundary_common_module
+#endif
+
+    public :: bc_buffers, bcxb, bcxe, bcyb, bcye, bczb, bcze
+
+#ifdef MFC_MPI
+    public :: MPI_BC_TYPE_TYPE, MPI_BC_BUFFER_TYPE
+#endif
+
+contains
+
+    subroutine s_initialize_boundary_common_module()
+
+#ifdef MFC_PRE_PROCESS
+        allocate (bc_buffers(1:num_dims, -1:1))
+
+        allocate (bc_buffers(1, -1)%sf(1:sys_size, 0:n, 0:p))
+        allocate (bc_buffers(1, 1)%sf(1:sys_size, 0:n, 0:p))
+        if (n > 0) then
+            allocate (bc_buffers(2, -1)%sf(-buff_size:m + buff_size, 1:sys_size, 0:p))
+            allocate (bc_buffers(2, 1)%sf(-buff_size:m + buff_size, 1:sys_size, 0:p))
+            if (p > 0) then
+                allocate (bc_buffers(3, -1)%sf(-buff_size:m + buff_size, -buff_size:n + buff_size, 1:sys_size))
+                allocate (bc_buffers(3, 1)%sf(-buff_size:m + buff_size, -buff_size:n + buff_size, 1:sys_size))
+            end if
+        end if
+#endif
+
+#ifdef MFC_SIMULATION
+        bcxb = bc_x%beg; bcxe = bc_x%end; bcyb = bc_y%beg; bcye = bc_y%end; bczb = bc_z%beg; bcze = bc_z%end
+
+        @:ALLOCATE(bc_buffers(1:num_dims, -1:1))
+
+        @:ALLOCATE(bc_buffers(1, -1)%sf(1:sys_size, 0:n, 0:p))
+        @:ALLOCATE(bc_buffers(1, 1)%sf(1:sys_size, 0:n, 0:p))
+        @:ACC_SETUP_SFs(bc_buffers(1,-1), bc_buffers(1,1))
+        if (n > 0) then
+            @:ALLOCATE(bc_buffers(2,-1)%sf(-buff_size:m+buff_size,1:sys_size,0:p))
+            @:ALLOCATE(bc_buffers(2,1)%sf(-buff_size:m+buff_size,1:sys_size,0:p))
+            @:ACC_SETUP_SFs(bc_buffers(2,-1), bc_buffers(2,1))
+            if (p > 0) then
+                @:ALLOCATE(bc_buffers(3,-1)%sf(-buff_size:m+buff_size,-buff_size:n+buff_size,1:sys_size))
+                @:ALLOCATE(bc_buffers(3,1)%sf(-buff_size:m+buff_size,-buff_size:n+buff_size,1:sys_size))
+                @:ACC_SETUP_SFs(bc_buffers(3,-1), bc_buffers(3,1))
+            end if
+        end if
+#endif
```

Comment:
I think you can remove the `#ifdef MFC_PRE_PROCESS` copy of these allocations, and discard the `#ifdef MFC_SIMULATION` surrounding these allocations. The `@:ALLOCATE` macro works in pre-process. Also, these should only be allocated if they're needed, ie if there's dirichlet BCs.